charm/src/charm.py

```diff
@@ -30,63 +35,55 @@ def __init__(self, *args: typing.Any) -> None:
             args: passthrough to CharmBase.
         """
         super().__init__(*args)
+        self.dns_record = dns_record.DNSRecordRequires(self)
         self.actions_observer = actions.Observer(self)
-        self.framework.observe(self.on.collect_app_status, self._on_collect_app_status)
-
-    def _on_config_changed(self, _event: ops.ConfigChangedEvent) -> None:
-        """Config changed handler.
-
-        Args:
-            _event: the event triggering the handler.
-        """
-        self._copy_files()
-        super()._on_config_changed(_event)
+        self.framework.observe(
+            self.on[CONTAINER_NAME].pebble_custom_notice, self._on_pebble_custom_notice
+        )
 
-    def _on_pebble_ready(self, _event: ops.PebbleReadyEvent) -> None:
-        """Pebble ready handler.
+    def _on_pebble_custom_notice(self, event: ops.PebbleCustomNoticeEvent) -> None:
+        """Handle pebble custom notice event.
 
         Args:
-            _event: the event triggering the handler.
+            event: Pebble custom notice event.
         """
-        self._copy_files()
-        super()._on_pebble_ready(_event)
-
-    def _copy_files(self) -> None:
-        """Copy files needed by git."""
-        if not self._container.can_connect():
-            return
-        if not self.config.get("git-repo") or not self.config.get("git-ssh-key"):
+        entries = []
+        known_notice = False
+        if event.notice.key.startswith("dns.local/write"):
+            fqdn = event.notice.last_data["fqdn"].strip("'").strip()
+            rdata = event.notice.last_data["rdata"].strip("'").strip()
+            entry = dns_record.RequirerEntry(
+                host_label=fqdn.split(".")[0],
+                domain=fqdn.split(".", 1)[1],
+                ttl=600,
+                record_class="IN",
+                record_type="TXT",
+                record_data=rdata,
+                uuid=uuid.uuid5(UUID_NAMESPACE, f"{fqdn} {rdata}"),
+            )
+            logger.debug("DNS record request: %s", entry)
+            entries = [entry]
+            known_notice = True
+
+        if event.notice.key.startswith("dns.local/remove"):
+            # For now, we remove everything
+            # since the default is no entries, we have nothing to do here.
+            known_notice = True
```

Comment:
There will be multiple requests in parallel, so we need to add / remove based on the FQDN and not override the whole databag when a new request comes

Reply:
Using the new lib, I did so that only the entries with a corresponding FQDN to the arriving notice are removed